google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkCommitBatch.java

```diff
@@ -16,32 +16,164 @@
 
 package com.google.cloud.firestore;
 
+import com.google.api.core.ApiAsyncFunction;
+import com.google.api.core.ApiFunction;
 import com.google.api.core.ApiFuture;
+import com.google.api.core.ApiFutures;
+import com.google.api.core.SettableApiFuture;
+import com.google.cloud.Timestamp;
 import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.google.firestore.v1.BatchWriteRequest;
+import com.google.firestore.v1.BatchWriteResponse;
+import io.grpc.Status;
+import io.opencensus.trace.AttributeValue;
+import io.opencensus.trace.Tracing;
+import java.util.ArrayList;
+import java.util.List;
+import javax.annotation.Nullable;
 
 /** Used to represent a batch on the BatchQueue. */
 class BulkCommitBatch extends UpdateBuilder<ApiFuture<WriteResult>> {
+  /**
+   * Used to represent the state of batch.
+   *
+   * <p>Writes can only be added while the batch is OPEN. For a batch to be sent, the batch must be
+   * READY_TO_SEND. After a batch is sent, it is marked as SENT.
+   */
+  enum BatchState {
+    OPEN,
+    READY_TO_SEND,
+    SENT,
+  }
+
+  protected BatchState state = BatchState.OPEN;
 
   BulkCommitBatch(FirestoreImpl firestore, int maxBatchSize) {
     super(firestore, maxBatchSize);
   }
 
-  BulkCommitBatch(FirestoreImpl firestore, BulkCommitBatch retryBatch) {
-    super(firestore);
+  ApiFuture<WriteResult> wrapResult(DocumentReference documentReference) {
+    return processLastOperation(documentReference);
+  }
+
+  /**
+   * Commits all pending operations to the database and verifies all preconditions.
+   *
+   * <p>The writes in the batch are not applied atomically and can be applied out of order.
+   */
+  ApiFuture<List<BatchWriteResult>> bulkCommit() {
```

Comment:
I tried extracting a `commitHelper()` from `bulkCommit()` and `commit()`, but couldn't figure out how to properly parameterize everything.

Aside from reducing code duplication, the main reason I want to do this is because moving this over requires a getter for `writes` and exposing `committed` as `protected`. If we could move the logic for these parts into `UpdateBuilder`, we could make both of those variables private.